enhancements/installer/assisted-installer.md

```diff
@@ -0,0 +1,349 @@
+bare metal, with as little up-front infrastructure dependencies as
+possible. This person has access to servers and wants to run workloads
+quickly. They do not necessarily have the administrative privileges to
+create private VLANs, configure DHCP/PXE servers, or manage other
+aspects of the infrastructure surrounding the hardware where the
+cluster will run. If they do have the required privileges, they may
+not want to delegate them to the OpenShift installer for an IPI
+installation, preferring instead to use their existing tools and
+processes for some or all of that configuration. They are willing to
+accept that the cluster they build may not have all of the same
+features as an IPI cluster on day 1, but that by taking additional
+steps later they will be able to add more automatic infrastructure
+management and make it fully production-ready.
+
+### Goals
+
+- Make initial deployment of usable and supportable clusters simpler.
+- Move more infrastructure configuration from day 1 to day 2.
+- Support connected and disconnected on-premise deployments.
+- Support existing IPI features, especially for day 2 cluster
+  management and scale-out.
+
+### Non-Goals
+
+- Because the initial focus is for bare metal, this enhancement does
+  not exhaustively cover variations needed to offer similar features
+  on other platforms (such as changes to image formats, the way a host
+  boots, etc.). It is desirable to support those platforms, but that
+  work will be described separately.
+- Replace the existing OpenShift installer.
+- Describe how these workflows would work for multi-cluster
+  deployments managed with Hive or ACM.
+
+## Proposal
+
+There are several separate changes to enable the assisted installer
+workflows, including a GUI front-end for the installer, a cloud-based
+orchestration service, and changes to the installer and bootstrapping
+process.
+
+### Connected Deployments
+
+The process starts when the user goes to an "assisted installer"
+application running on `cloud.redhat.com`, enters details needed by
+the installer (OpenShift version, ssh keys, proxy settings, etc.), and
+then downloads a live RHCOS ISO image with the software and settings
+they need to complete the installation locally.
+
+The user then boots the live ISO on each host they want to be part of
+the cluster (control plane and workers). They can do this by hand
+using thumb drives, by attaching the ISO using virtual media support
+in the BMC of the host, or any other way they choose.
+
+When the ISO boots, it starts an agent that communicates with the REST
+API for the assisted installer service running on `cloud.redhat.com`
+to receive instructions. The agent identifies itself based on the
+settings in the ISO and hardware details it discovers from the host it
+is running on. Communication always flows from agent to service (using
+polling, long polling, web sockets, or some other similar technique)
+so that firewalls and proxies work as expected.
```

Comment:
This I imagine will be an area of intense interest - i.e. there will be a desire to closely examine how we're going about having a cloud service control software installation in a private data center. Details like how the agent authenticates the service, and what instructions the agent will accept from service.

Reply:
This is a good point, particularly if it reuses the pull secret for auth, as that's so far been used only to control access when pulling images, and not as a token to permit mutating system state - I wonder if that is an argument to never run the agent after the initial installation?

Reply:
@avishayt or @romfreiman can you suggest some additional text for this section?

Reply:
I'll have a go at it...:

When the ISO boots, it starts an agent that communicates with the REST API for the assisted installer service running on `cloud.redhat.com` to receive instructions. The agent registers the host with the service, using the user's pull secret embedded in the ISO's ignition to authenticate. The agent identifies itself based on the serial number from the host it's running on. Communication always flows from agent to service via HTTPS so that firewalls and proxies work as expected.

Each host agent periodically asks the service what tasks to perform, and the service replies with a list of commands and arguments.  A command can be to:
1. Return hardware information for its host
2. Return L2 and L3 connectivity information between its host and the other hosts (the IPs and MAC addresses of the other hosts are passed as arguments)
3. Begin the installation of its host (arguments include the host's role, boot device, etc.).  The agent executes different installation logic depending on its role (bootstrap-master, master, or worker).

The agent posts the results for the command back to the service.  During the actual installation, the agents post progress.

Reply:
I've replaced the paragraph with @avishayt's text. That broke the link to this comment thread, so for reference it's now on lines 136-158.